lib/Conversion/GPUToROCM/ConvertKernelFuncToHSACO.cpp

```diff
@@ -0,0 +1,408 @@
+
+    // This is a module containing a GPU kernel function, we have work to do!
+
+    // Make sure the AMDGPU target is initialized.
+    LLVMInitializeAMDGPUTarget();
+    LLVMInitializeAMDGPUTargetInfo();
+    LLVMInitializeAMDGPUTargetMC();
+    LLVMInitializeAMDGPUAsmPrinter();
+
+    auto llvmModule = translateModuleToROCDLIR(getModule());
+    if (!llvmModule) {
+      return signalPassFailure();
+    }
+
+    for (auto function : getModule().getOps<FuncOp>()) {
+      if (!gpu::GPUDialect::isKernel(function)) {
+        continue;
+      }
+
+      if (failed(translateGpuKernelToHSACOAnnotation(*llvmModule, function)))
+        signalPassFailure();
+    }
+
+    // dump the MLIR
+    if (getenv("MLIR_ROCM_DUMP_AFTER_CONV_GPU_KERNEL")) {
+      std::cerr << "MLIR - After call to convert GPU Kernel" << std::endl;
+      getModule().dump();
+    }
+  }
+
+private:
+  LogicalResult translateGpuKernelToHSACOAnnotation(llvm::Module &llvmModule,
+                                                    FuncOp &function);
+
+  OwnedHSACO convertModuleToHSACO(llvm::Module &llvmModule, FuncOp &function);
+
+  OwnedHSACO emitModuleToHSACO(llvm::Module &llvmModule,
+                               llvm::TargetMachine &targetMachine);
+
+  OwnedHSACO emitModuleToHSACOForTesting(llvm::Module &llvmModule,
+                                         FuncOp &function);
+
+  rocm::HSACOGeneratorConfig config;
+};
+
+// get the "-mcpu" option string corresponding to the given AMDGPU version enum
+std::string GetMcpuOptionString(rocm::AMDGPUVersion v) {
+  switch (v) {
+  case rocm::AMDGPUVersion::GFX900:
+    return "gfx900";
+    break;
+  }
+  return "<invalid AMDGPU version>";
+}
+
+// get filename for file containing the AMDGPU version specific bitcodes
+std::string GetBitcodeFilename(rocm::AMDGPUVersion v) {
+  switch (v) {
+  case rocm::AMDGPUVersion::GFX900:
+    return "oclc_isa_version_900.amdgcn.bc";
+    break;
+  }
+  return "<invalid AMDGPU version>";
+}
+
+// get the option string corresponding to the given HSACO version enum
+std::string GetCodeObjectOptionString(rocm::HSACOVersion v) {
+  switch (v) {
+  case rocm::HSACOVersion::V3:
+    return "-code-object-v3";
+    break;
+  }
+  return "invalid HSACO version";
+}
+
+// Gets the ROCm-Device-Libs filenames for a particular AMDGPU version.
+static std::vector<std::string>
+GetROCDLPaths(const rocm::AMDGPUVersion amdgpuVersion,
+              const std::string &rocdlDir) {
+
+  // AMDGPU version-neutral bitcodes.
+  static std::vector<std::string> *rocdlFilenames =
+      new std::vector<std::string>({"hc.amdgcn.bc", "opencl.amdgcn.bc",
+                                    "ocml.amdgcn.bc", "ockl.amdgcn.bc",
+                                    "oclc_finite_only_off.amdgcn.bc",
+                                    "oclc_daz_opt_off.amdgcn.bc",
+                                    "oclc_correctly_rounded_sqrt_on.amdgcn.bc",
+                                    "oclc_unsafe_math_off.amdgcn.bc"});
```

Comment:
@deven-amd need to add `oclc_wavefrontsize64_on.amdgcn.bc`. refer to:

https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/pull/590/files

Reply:
done